routing/src/vrf.rs

```diff
@@ -0,0 +1,447 @@
+
+/// Every VRF is univocally identified with a numerical VRF id
+pub type VrfId = u32;
+
+#[derive(Debug, Clone, Eq, Hash, PartialEq)]
+#[doc = "A next-hop in the VRF"]
+pub struct RouteNhop {
+    pub vrfid: VrfId,
+    pub key: NhopKey,
+}
+impl Default for RouteNhop {
+    fn default() -> Self {
+        Self {
+            vrfid: 0,
+            key: NhopKey::with_drop(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Route {
+    pub rtype: RouteType,
+    pub distance: RouteDistance,
+    pub metric: RouteMetric,
+    pub s_nhops: Vec<ShimNhop>,
+}
+impl Default for Route {
+    fn default() -> Self {
+        Self {
+            rtype: RouteType::Other,
+            distance: 0,
+            metric: 0,
+            s_nhops: Vec::with_capacity(1),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct ShimNhop {
+    pub shim: VrfId,
+    pub rc: Rc<Nhop>,
+}
+
+#[allow(unused)]
+pub struct Vrf {
+    #[allow(unused)]
+    pub name: String,
+    pub vrfid: VrfId,
+    pub(crate) routesv4: RTrieMap<Ipv4Prefix, Route>,
+    pub(crate) routesv6: RTrieMap<Ipv6Prefix, Route>,
+    pub(crate) nhstore: NhopStore,
+    pub(crate) vni: Option<Vni>,
+}
+
+#[allow(dead_code)]
+impl Vrf {
+    /////////////////////////////////////////////////////////////////////////
+    /// Create a new VRF with the given name and vrfId
+    /// Initial capacities are 0. We may want to have some default sizes.
+    /// Else, we can always call Self::with_capacities().
+    /////////////////////////////////////////////////////////////////////////
+    pub fn new(name: &str, vrfid: VrfId) -> Self {
+        Self::with_capacities(name, vrfid, 0, 0)
+    }
+
+    /////////////////////////////////////////////////////////////////////////
+    /// Create VRF with some initial capacities
+    /////////////////////////////////////////////////////////////////////////
+    pub fn with_capacities(name: &str, vrfid: VrfId, capa_v4: usize, capa_v6: usize) -> Self {
+        let mut vrf = Self {
+            name: name.to_owned(),
+            vrfid,
+            routesv4: RTrieMap::with_capacity(capa_v4),
+            routesv6: RTrieMap::with_capacity(capa_v6),
+            nhstore: NhopStore::new(),
+            vni: None,
+        };
+        /* add default routes with default next-hop with action DROP */
+        vrf.add_route(
+            &Prefix::root_v4(),
+            Route::default(),
+            &[RouteNhop::default()],
+        );
+        vrf.add_route(
+            &Prefix::root_v6(),
+            Route::default(),
+            &[RouteNhop::default()],
+        );
+        vrf
+    }
+
+    /////////////////////////////////////////////////////////////////////////
+    /// Set the VNI for a Vrf
+    /////////////////////////////////////////////////////////////////////////
+    pub fn set_vni(&mut self, vni: Vni) {
+        self.vni = Some(vni);
+    }
+
+    #[inline(always)]
+    #[must_use]
```

Comment:
`inline` annotations are almost never necessary if you are using LTO (as we are).

I recommend avoiding them unless we need to bully the compiler for some reason.

Reply:
Happy to remove inline if you think it's not necessary. Note however, that the inline functions are not pub and only called from the ip-version agnostic counterparts (I did that split to keep the code simple).  So, providing that hint to the compiler may be okay ?

Reply:
Does the compiler even use the hint?  I know in C/C++ it often ignores it in favor of its own heuristic, so I'm curious if this is really honored in Rust.

Reply:
In my experience the compiler will do its best to honor `#[inline(always)]`

That said, the old joke is that LLVM's heuristic for inlining functions is "yes"

Reply:
In my experience the compiler will do its best to honor `#[inline(always)]`.

The annotation is generally most useful when you have a performance sensitive function being called across crate boundaries and you don't want LTO for some reason.  rustc uses "thin local" lto by default which basically means LTO within a single crate.  If you turn that to "thin" or "fat" then you get LTO across all crates in the compile job.

As for "will LLVM inline this," the old joke is that LLVM's heuristic for deciding if it should inline a function is "yes"

Still, @Fredi-raspall is correct that the annotation is harmless here.